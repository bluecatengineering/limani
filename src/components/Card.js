/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import PropTypes from 'prop-types';
import { Layer } from '@bluecateng/pelagos';
import { ArrowRight } from '@carbon/icons-react';
import './Card.less';

/**
 * Card component presents a rectangle box layout
 * with some data to fill it up with.
 */

const Card = ({ title, href, description, className }) => {
    return (
        <Layer className={`Card${className ? ' ' + className : ''}`}>
            <a className='Card__link' href={href}>
                <div>
                    <span className='Card__title'>{title}</span>
                    {description && (
                        <span className='Card__description'>{description}</span>
                    )}
                </div>
                <div className='Card__iconContainer'>
                    <ArrowRight size={20} className='Card__arrowRight' />
                </div>
            </a>
        </Layer>
    );
};

Card.propTypes = {
    /** Title for the card */
    title: PropTypes.string.isRequired,

    /** Link is attached to the href attribute for the card */
    href: PropTypes.string.isRequired,

    /** Description inside in the card */
    description: PropTypes.string,

    /** The component class name(s). */
    className: PropTypes.string,
};

export default Card;
