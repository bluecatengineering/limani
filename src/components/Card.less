// Copyright 2023 BlueCat Networks Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
@import '~@bluecateng/pelagos/less/spacing';
@import '~@bluecateng/pelagos/less/fonts';

.Card {
    display: flex;
    flex-direction: column;
    gap: @sp-08;
    background-color: var(--layer-01);
    width: 272px;
    height: 150px;
    padding: @sp-16;
    text-decoration: none;

    &__title {
        color: var(--link-primary);
        font-size: @sp-16;
        line-height: @sp-32;
        width: 184px;
        text-align: left;
        text-decoration: underline;
    }

    &__description {
        display: -webkit-box;
        -webkit-line-clamp: 2;
        -webkit-box-orient: vertical;
        color: var(--text-placeholder);
        font-size: @sp-12;
        line-height: @sp-16;
        max-width: 235px;
        text-align: left;
        text-overflow: ellipsis;
        overflow: hidden;
    }

    &:hover {
        background-color: var(--layer-hover-01);
        outline: 1.5px solid var(--border-interactive);
        box-shadow: 0 0 5px @sp-02 var(--border-subtle-selected-01);
        text-decoration: none;
        .Card__title {
            color: var(--link-primary-hover);
        }
    }
    &:active {
        background-color: var(--layer-active-01);
    }

    &__path {
        display: block;
        color: var(--text-primary);
        font-size: @sp-16;
        line-height: @sp-24;
        text-align: left;
    }

    &__pathTitle {
        display: block;
        color: var(--text-secondary);
        font-size: @sp-12;
        line-height: @sp-16;
        text-align: left;
    }
}
