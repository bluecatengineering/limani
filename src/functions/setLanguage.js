/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import l10n from '../l10n';

const loaders = {
    en: () => import('../l10n/en.po'),
    zz: () => import('../l10n/zz.po'),
};

/**
 * Sets the language for Limani. The default language is English, there is no
 * need to call this function unless the selected language is different.
 * This function should be called as early as possible,
 * the caller must catch the returned promise and handle any errors.
 * @param {'en'|'es'} language The language translated content to load for.
 * @returns {*|Promise<never>}
 *
 * @example
 * import {setLanguage} from '@bluecateng/limani';
 *
 * setLanguage('es').catch((error) => {
 *     // handle error
 * })
 */
const setLanguage = (language) => {
    const loader = loaders[language];
    return loader
        ? loader().then(({ default: data }) => l10n.load(data))
        : Promise.resolve(console.warn(`Unknown language ${language}`));
};

export default setLanguage;
