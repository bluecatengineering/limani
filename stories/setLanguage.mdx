{/* Copyright 2023 BlueCat Networks Inc */}
{/*  */}
{/* Permission is hereby granted, free of charge, to any person obtaining a copy */}
{/* of this software and associated documentation files (the "Software"), to deal */}
{/* in the Software without restriction, including without limitation the rights */}
{/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell */}
{/* copies of the Software, and to permit persons to whom the Software is */}
{/* furnished to do so, subject to the following conditions: */}
{/*  */}
{/* The above copyright notice and this permission notice shall be included in all */}
{/* copies or substantial portions of the Software. */}
{/*  */}
{/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR */}
{/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, */}
{/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE */}
{/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER */}
{/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, */}
{/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE */}
{/* SOFTWARE. */}
import { Meta } from '@storybook/addon-docs';

<Meta title='functions/setLanguage' />

# setLanguage

Sets the language for Limani. The default language is set to English i.e. `en`.
There is no need to call this function unless the selected
language is other than `en`. This function should be called as early as
possible, and wait before rendering the components,as it's an asynchronous operation.
Currently only `en` is supported.

## Parameters

<table>
    <thead>
        <tr>
            <th>Param</th>
            <th>Type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>language</td>
            <td>`String`</td>
            <td>The language that needs to be set.</td>
        </tr>
    </tbody>
</table>

## Returns

`Promise` - Resolves successfully on language change.

## Example

```js
import { setLanguage } from '@bluecateng/limani';

const Example = () => (
    setLanguage('es').catch((error) => {
        // handle error
    })
    ...
);
```
