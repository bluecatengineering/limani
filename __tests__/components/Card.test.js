/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { shallow } from 'enzyme';
import Card from '../../src/components/Card';
import setLanguage from '../../src/functions/setLanguage';

jest.unmock('../../src/components/Card');

describe('Card', () => {
    describe('Rendering', () => {
        it('Render the Card component with default props', () => {
            const wrapper = shallow(<Card title='Add Host record' />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('Render the Card component with props', () => {
            const wrapper = shallow(
                <Card
                    href='#'
                    title='Add Host Record'
                    description='Add a host record on BAM'
                    path={['Host Record']}
                    className='test_cn'
                />,
            );
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('Render the Card component with pseudo translated text', () => {
            setLanguage('zz').finally(() => {
                const wrapper = shallow(
                    <Card
                        href='#'
                        title='Add Host Record'
                        description='Add a host record on BAM'
                        path={['Host Record']}
                        className='test_cn'
                    />,
                );
                expect(wrapper.getElement()).toMatchSnapshot();
            });
        });
    });
});
