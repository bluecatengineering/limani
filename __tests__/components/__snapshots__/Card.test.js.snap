// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Rendering Render the Card component with default props 1`] = `
<Layer
  className="Card"
>
  <a
    className="Card__link"
  >
    <div>
      <span
        className="Card__title"
      >
        Add Host record
      </span>
    </div>
    <div
      className="Card__iconContainer"
    >
      <ArrowRight
        className="Card__arrowRight"
        size={20}
      />
    </div>
  </a>
</Layer>
`;

exports[`Card Rendering Render the Card component with props 1`] = `
<Layer
  className="Card test_cn"
>
  <a
    className="Card__link"
    href="#"
  >
    <div>
      <span
        className="Card__title"
      >
        Add Host Record
      </span>
      <span
        className="Card__description"
      >
        Add a host record on BAM
      </span>
    </div>
    <div
      className="Card__iconContainer"
    >
      <ArrowRight
        className="Card__arrowRight"
        size={20}
      />
    </div>
  </a>
</Layer>
`;

exports[`Card Rendering Render the Card component with pseudo translated text 1`] = `
<Layer
  className="Card test_cn"
>
  <a
    className="Card__link"
    href="#"
  >
    <div>
      <span
        className="Card__title"
      >
        Add Host Record
      </span>
      <span
        className="Card__description"
      >
        Add a host record on BAM
      </span>
    </div>
    <div
      className="Card__iconContainer"
    >
      <ArrowRight
        className="Card__arrowRight"
        size={20}
      />
    </div>
  </a>
</Layer>
`;
